psi4/src/psi4/dlpno/dlpno.cc

```diff
@@ -0,0 +1,1243 @@
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "dlpno.h"
+#include "sparse.h"
+
+#include "psi4/lib3index/3index.h"
+#include "psi4/libdiis/diismanager.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/libmints/local.h"
+#include "psi4/libmints/matrix.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/orthog.h"
+#include "psi4/libmints/twobody.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libpsi4util/PsiOutStream.h"
+#include "psi4/libpsi4util/process.h"
+#include "psi4/libqt/qt.h"
+
+#include <algorithm>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace psi {
+namespace dlpno {
+
+DLPNO::DLPNO(SharedWavefunction ref_wfn, Options& options) : Wavefunction(options) {
+    shallow_copy(ref_wfn);
+    reference_wavefunction_ = ref_wfn;
+
+    common_init();
+}
+DLPNO::~DLPNO() {}
+void DLPNO::common_init() {
+    print_ = options_.get_int("PRINT");
+    debug_ = options_.get_int("DEBUG");
+
+    T_CUT_PNO_ = options_.get_double("T_CUT_PNO");
+    T_CUT_DO_ = options_.get_double("T_CUT_DO");
+
+    if (options_.get_str("DLPNO_ALGORITHM") == "MP2") {
+        algorithm_ = MP2;
+    } else {
+        throw PSIEXCEPTION("Requested DLPNO algorithm has NOT been implemented yet");
+    }
+
+    // did the user manually change expert level options?
+    bool T_CUT_PNO_changed = options_["T_CUT_PNO"].has_changed();
+    bool T_CUT_DO_changed = options_["T_CUT_DO"].has_changed();
+
+    // if not, values are determined by the user-friendly "PNO_CONVERGENCE"
+    if (algorithm_ == MP2) {
+        if (options_.get_str("PNO_CONVERGENCE") == "LOOSE") {
+            if (!T_CUT_PNO_changed) T_CUT_PNO_ = 1e-7;
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 2e-2;
+        } else if (options_.get_str("PNO_CONVERGENCE") == "NORMAL") {
+            if (!T_CUT_PNO_changed) T_CUT_PNO_ = 1e-8;
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 1e-2;
+        } else if (options_.get_str("PNO_CONVERGENCE") == "TIGHT") {
+            if (!T_CUT_PNO_changed) T_CUT_PNO_ = 1e-9;
+            if (!T_CUT_DO_changed) T_CUT_DO_ = 5e-3;
+        }
+    } else {
+        throw PSIEXCEPTION("Requested DLPNO algorithm has NOT been implemented yet");
```

Comment:
Including the wrapper, this is the third check on the allowed algorithm, plus the enum, plus read_options only lets it be set to MP2. Checks are good, but you might want to reassess the caution/clutter balance once all the code rearrangement settles down.

Reply:
Yes, in this case, this check is redundant